<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Rcollectl: instrumentation for R processes on linux systems • Rcollectl</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Rcollectl: instrumentation for R processes on linux systems">
<meta property="og:description" content="Provide functions to obtain instrumentation data on
    processes in a unix environment.  Parse output of a collectl
    run.  Vizualize aspects of system usage.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Rcollectl</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/Rcollectl.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/Rcollectl">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<!--
[![BioC
status](http://www.bioconductor.org/shields/build/release/bioc/Rcollectl.svg)](https://bioconductor.org/checkResults/release/bioc-LATEST/Rcollectl)
[![BioC dev
status](http://www.bioconductor.org/shields/build/devel/bioc/Rcollectl.svg)](https://bioconductor.org/checkResults/devel/bioc-LATEST/Rcollectl)
-->
<p><a href="https://github.com/vjcitn/Rcollectl/actions"><img src="https://github.com/vjcitn/Rcollectl/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a> <a href="https://codecov.io/gh/vjcitn/Rcollectl?branch=main"><img src="https://codecov.io/gh/vjcitn/Rcollectl/branch/main/graph/badge.svg" alt="Codecov test coverage"></a> <!--
[![Support site activity, last 6 months: tagged questions/avg. answers
per question/avg. comments per question/accepted answers, or 0 if no
tagged
posts.](http://www.bioconductor.org/shields/posts/Rcollectl.svg)](https://support.bioconductor.org/t/Rcollectl/)
--> <a href="https://github.com/vjcitn/Rcollectl/issues"><img src="https://img.shields.io/github/issues/vjcitn/Rcollectl" alt="GitHub issues"></a> <!-- badges: end --></p>
<div id="rcollectl-simple-interfaces-to-collectl-output" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#rcollectl-simple-interfaces-to-collectl-output" class="anchor"></a>Rcollectl: simple interfaces to collectl output</h1></div>
<p>Profiling R computations is supported by Rprof and profvis. These measure memory consumption and function execution counts and timings. In workflow design we are also interested in CPU load, disk I/O, and network traffic. There is no standard portable approach to measuring these. In this package, we focus on linux systems measurement with the <a href="http://collectl.sourceforge.net/index.html">collectl</a> suite of tools.</p>
<div id="notes-from-the-anvil-workspace-description" class="section level2">
<h2 class="hasAnchor">
<a href="#notes-from-the-anvil-workspace-description" class="anchor"></a>Notes from the AnVIL workspace description</h2>
<p>Our objective in this workspace is demonstration of instrumentation of bioinformatic tasks. By “instrumentation” we mean the definition and use of tools that measure computational resource consumption: memory, disk, CPU, network traffic.</p>
<p>(On AnVIL, we do not know how to identify the network device and data on network traffic are not obtained as of 1/1/2021.)</p>
<p>This Dashboard Description discusses details of an example using <a href="https://github.com/COMBINE-lab/salmon">salmon</a> to carry out quantification of an RNA-seq experiment.</p>
</div>
<div id="basic-resources" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-resources" class="anchor"></a>Basic resources</h2>
<div id="workspace-backbone" class="section level3">
<h3 class="hasAnchor">
<a href="#workspace-backbone" class="anchor"></a>Workspace backbone</h3>
<p>This workspace was initialized using AnVILPublish. The runtime/cloud environment used as of 2020-01-04 is</p>
<p>dockerhub vjcitn/instr:0.0.3</p>
</div>
<div id="raw-data" class="section level3">
<h3 class="hasAnchor">
<a href="#raw-data" class="anchor"></a>Raw data</h3>
<p>The FASTQ files underlying the “airway” RNA-seq workflow were collected using commands like</p>
<pre><code>gsutil -u landmarkanvil2 ls  gs://sra-pub-run-3/SRR1039512/SRR1039512.1
gsutil -u landmarkanvil2 cp  gs://sra-pub-run-4/SRR1039513/SRR1039513.1 .
gsutil -u landmarkanvil2 cp  gs://sra-pub-run-4/SRR1039516/SRR1039516.1 .
gsutil -u landmarkanvil2 cp  gs://sra-pub-run-2/SRR1039517/SRR1039517.1 .
gsutil -u landmarkanvil2 cp  gs://sra-pub-run-2/SRR1039520/SRR1039520.1 .
gsutil -u landmarkanvil2 cp  gs://sra-pub-run-3/SRR1039521/SRR1039521.1 .</code></pre>
<p>A public bucket with the extracted fastq was then produced via</p>
<pre><code>fastq-dump --split-3 *512.1 &amp;
fastq-dump --split-3 *513.1 &amp;
fastq-dump --split-3 *516.1 &amp;
fastq-dump --split-3 *517.1 &amp;
fastq-dump --split-3 *520.1 &amp;
fastq-dump --split-3 *521.1 &amp;
gzip *fastq
gsutil ls gs://bioc-airway-fastq
gsutil cp SRR1039508.1_2.fastq.gz gs://bioc-airway-fastq/</code></pre>
</div>
</div>
<div id="quantification-process" class="section level2">
<h2 class="hasAnchor">
<a href="#quantification-process" class="anchor"></a>Quantification process</h2>
<div id="software-and-indexing-resources" class="section level3">
<h3 class="hasAnchor">
<a href="#software-and-indexing-resources" class="anchor"></a>Software and indexing resources</h3>
<p>Our aim is to analyze the resource consumption in processing this sort of data with R using Rcollectl. We’ll describe how to run salmon to generate quantifications,and we will attempt to give estimates for processing various volumes of data by various approaches.</p>
<p>We installed snakemake with pip3 and salmon via</p>
<pre><code>wget https://github.com/COMBINE-lab/salmon/releases/download/v1.4.0/salmon-1.4.0_linux_x86_64.tar.gz</code></pre>
<p>We obtained GENCODE transcript sequences:</p>
<pre><code>wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_36/gencode.v36.transcripts.fa.gz</code></pre>
<p>And built the index <code>gencode.v36_salmon_1.4.0</code> via</p>
<pre><code>salmon index --gencode -t gencode.v36.transcripts.fa.gz -i gencode.v36_salmon_1.4.0</code></pre>
</div>
<div id="snakemake-for-workflow-definition" class="section level3">
<h3 class="hasAnchor">
<a href="#snakemake-for-workflow-definition" class="anchor"></a>Snakemake for workflow definition</h3>
<p>For a single sample (paired-end) we have a snakemake file:</p>
<pre><code>rstudio@saturn-1f2f18e5-4182-40c0-8449-1f301b5c3b03:~$ cat snakemake_one
DATASETS = ["SRR1039508.1"]

SALMON = "/home/rstudio/bin/salmon"

rule all:
  input: expand("quants/{dataset}/quant.sf", dataset=DATASETS)

rule salmon_quant:
    input:
        r1 = "fastq/{sample}_1.fastq.gz",
        r2 = "fastq/{sample}_2.fastq.gz",
        index = "/home/rstudio/gencode.v36_salmon_1.4.0"
    output:
        "quants/{sample}/quant.sf"
    params:
        dir = "quants/{sample}"
    shell:
        "{SALMON} quant -i {input.index} -l A -p 6 --validateMappings \
         --gcBias --numGibbsSamples 20 -o {params.dir} \
         -1 {input.r1} -2 {input.r2}"</code></pre>
<p>When used with</p>
<pre><code>snakemake -j1 --snakefile snakemake_one</code></pre>
<p>it takes 6 minutes to produce a folder <code>quants</code> with content</p>
<pre><code>└── SRR1039508.1
    ├── aux_info
    │   ├── ambig_info.tsv
    │   ├── bootstrap
    │   │   ├── bootstraps.gz
    │   │   └── names.tsv.gz
    │   ├── expected_bias.gz
    │   ├── exp_gc.gz
    │   ├── fld.gz
    │   ├── meta_info.json
    │   ├── observed_bias_3p.gz
    │   ├── observed_bias.gz
    │   └── obs_gc.gz
    ├── cmd_info.json
    ├── lib_format_counts.json
    ├── libParams
    │   └── flenDist.txt
    ├── logs
    │   └── salmon_quant.log
    └── quant.sf</code></pre>
</div>
</div>
<div id="instrumentation-of-this-process" class="section level2">
<h2 class="hasAnchor">
<a href="#instrumentation-of-this-process" class="anchor"></a>Instrumentation of this process</h2>
<p>We use the Rcollectl package (which must be installed from github.com/vjcitn/Rcollectl) to monitor resource consumption while the snakemake process is running.</p>
<div id="single-sample-6-threads" class="section level3">
<h3 class="hasAnchor">
<a href="#single-sample-6-threads" class="anchor"></a>Single sample 6 threads</h3>
<p>Here is a display of the resource consumption for the single (paired-end) sample:</p>
<p><img src="https://storage.googleapis.com/bioc-anvil-images/airway_onesamp_collectl.png" alt="single sample usage profile"></p>
<p>That run uses a setting of <code>-p 6</code> for <code>salmon quant</code>, which allows the algorithm to use 6 threads.</p>
</div>
<div id="single-sample-12-threads" class="section level3">
<h3 class="hasAnchor">
<a href="#single-sample-12-threads" class="anchor"></a>Single sample 12 threads</h3>
<p>We used a 16 core machine, and raised the value of <code>-p</code> to 12 to obtain the following profile:</p>
<p><img src="https://storage.googleapis.com/bioc-anvil-images/salmon_p12_one.png" alt="single sample usage, 12 threads"></p>
</div>
<div id="eight-samples-12-threads" class="section level3">
<h3 class="hasAnchor">
<a href="#eight-samples-12-threads" class="anchor"></a>Eight samples 12 threads</h3>
<p><img src="https://storage.googleapis.com/bioc-anvil-images/salmon_p12_eight.png" alt="eight samples 12 threads"></p>
<p>It takes about 40 minutes to do the eight samples, using about 80% of the available CPUs and about 16 GB RAM overall.</p>
</div>
<div id="four-samples-3-threads-4-processes--j4-for-snakemake" class="section level3">
<h3 class="hasAnchor">
<a href="#four-samples-3-threads-4-processes--j4-for-snakemake" class="anchor"></a>Four samples 3 threads 4 processes (-j4 for snakemake)</h3>
<p>Finally, we obtain some data to compare thread-based parallelism to process parallelism. We’ll reduce the number of threads per sample, but run 4 samples at once via snakemake -j argument.</p>
<p><img src="https://storage.googleapis.com/bioc-anvil-images/salmon_p3_j4.png" alt="four samples 3 threads -j4"></p>
<p>It seems the throughput is a bit better here, but we should check the comparability of results of the different approaches before declaring this.</p>
</div>
</div>
<div id="conclusions" class="section level2">
<h2 class="hasAnchor">
<a href="#conclusions" class="anchor"></a>Conclusions</h2>
<p>Rcollectl can be used to obtain useful profiling information on AnVIL interactive work. We saw that the salmon quantification could be tuned to consume 80-90% of available CPUs. The RAM requirements and disk consumption are measurable. It does not seem possible to decouple the NCPU from the RAM elements in a fine-grained way in AnVIL interactive cloud environments. For 16 cores we could choose between 14.4 and 64 GB RAM. 64 GB seems too much for the task we are doing.</p>
<p>Additional work of interest: snakemake is used only because a working snakefile was published. We would like to be able to conduct these exercises in a more Bioconductor-native way. Specifically, we should have an object representing the read data in a self-describing way, and use BiocParallel to manage the execution approach.</p>
<p>Furthermore, as we define the self-describing representation of the read data in Bioconductor, we should consider how it maps to the AnVIL/Terra-native structures related to the workspace data tables, and DRS.</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/vjcitn/Rcollectl/">https://​github.com/​vjcitn/​Rcollectl/​</a>
</li>
<li>Report a bug at <br><a href="https://support.bioconductor.org/t/Rcollectl">https://​support.bioconductor.org/​t/​Rcollectl</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Vincent Carey <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-4046-0063" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#experimental"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
